### DISCUSSION TITLE
User provide part of the Jacobian and Hessian matrix

### DISCUSSION THREAD

**Discussion by shiftlin89** [casadi/casadi#3606] [Ideas]
Hi community members,

    My background is in power grid optimization. Usually, I need to solve a system with a size of over 10,000 buses. The number of nonlinear constraints is over 20,000 and the number of variables is over 30,000.
    
     It is still time-consuming to rely on the CADADi to generate Jacobian and Hessian.
    
    So I am wondering about the possibility that the user can provide part of the Jacobian and Hessian matrix so that it is still easy to model the problem and achieve fast computation.
    
    For example:
    
    cons1 = x^2;
    cons1_J=2*x;
    cons1_H=2;
    
    cons2 = y;
    
    opti.subject_to( cons1 <=0, cos1_J, cons1_H  ); //user provides part of the jacobian and hessian values
    opti.subject_to( cons2 <=0, [], []  );
    
    Thanks!

**Comment by sandeep026** [comment]
I am not aware of supplying a part of the hessian or Jacobian in the manner you have suggested. But there is provision to supply the enitre derivates yourself.

https://github.com/casadi/casadi/wiki/FAQ%3A-How-to-specify-a-custom-Hessian-approximation%3F

https://github.com/casadi/casadi/blob/main/docs/examples/matlab/sysid_gauss_newton.m

**Reply by shiftlin89** [reply]
Thank you!  Follow your suggestion, it is an idea that the user defines all the Jacobian and hessian matrices, in which part is provided by the user and part is calculated from the CASADi.

**Comment by jaeandersson** [comment]
You can provide parts of Jacobians or Hessians by reformulating your problem a bit.
You typically need to lift your formulation to a higher dimension, introducing new variables for the parts you want to provide manually. The new lifted variables can then be eliminated using the triangular solve node. Unfortunately I don't think there is more detailed description on how to do it in public sources.
